### ISSUE TITLE
64x32 panel with amperka (adafrult)

### ISSUE THREAD

**Issue by anp135** [hzeller/rpi-rgb-led-matrix#1100]
Hi. I have https://amperka.ru/product/rgb-led-matrix-driver-cap with some RGB matrix from Ali:
P4-2121-16S-HL 1.0 2P2751A 32x32
P4 256*128-2121A5 64x32
P5(2121)-16S-6432-B0 2PS1111F-HL 63x32
P6-3528-32*32-16S-HL1.1 2P2768A-HL 32x32

They are all upper half does not have blue, lower half does not have red. In the sabe time, when I use 
#./clock --led-cols=32 --led-rows=32 -d %H:%M:%S -f ../fonts/8x13B.bdf -b 30 -C 255,0,0 -B 255,255,255 --led-gpio-mapping=adafruit-hat

I see white background, in the upper half I see red clock and in the lower half I see blue the same clock.

For clock without background (-B), the clock line single and for upper half - does not have blue, for lower half does not have red.

Could you help?

**Comment by anp135** [comment]
[http://wiki.amperka.ru/_media/продукты:rgb-led-matrix-driver-cap:rgb-led-matrix-driver-cap_schematic.png?w=650&tok=645d9f](url)

**Comment by hzeller** [comment]
Do they use the same pin-mapping as this project ? If not, then you'd see some artifacts as missing colors.

Check their schematic and see what GPIO pin they connect to what output. Then add the mapping to here:
https://github.com/hzeller/rpi-rgb-led-matrix/blob/master/lib/hardware-mapping.c

Or maybe they use one of the existing already like `--led-gpio-mapping=adafruit-hat` or `--led-gpio-mapping=classic` ?

**Comment by anp135** [comment]
> Do they use the same pin-mapping as this project ? If not, then you'd see some artifacts as missing colors.
In examples they used adafruit-hat mapping, but I suppose something wrong.

> Check their schematic and see what GPIO pin they connect to what output. Then add the mapping to here:
> https://github.com/hzeller/rpi-rgb-led-matrix/blob/master/lib/hardware-mapping.c
Unfortunately I can't read schematic (link upper) for understood GPIO<->HUB75 connectivity -(.
 
> Or maybe they use one of the existing already like `--led-gpio-mapping=adafruit-hat` or `--led-gpio-mapping=classic` ?
--led-gpio-mapping=classic does not work completely (black matrix). I compiled with hardware adafruit-hat and this mapping a bit works:

./clock -f --led-cols=64 --led-rows=32 ../fonts/7x13.bdf -d "%H:%M:%S"  -b 40 -C 0,0,255 -B 255,0,0
upper half: red background, blue clock
down half: black background and red copy clock

./clock -f --led-cols=64 --led-rows=32 ../fonts/7x13.bdf -d "%H:%M:%S"  -b 40 -C 0,0,255 -B 0,0,255
upper half: completely black
footer half: completely blue

./clock -f --led-cols=64 --led-rows=32 ../fonts/7x13.bdf -d "%H:%M:%S"  -b 40 -C 0,255,0 -B 0,0,255
upper half: background black, green clock
footer half: completely blue

./clock -f --led-cols=64 --led-rows=32 ../fonts/7x13.bdf -d "%H:%M:%S"  -b 40 -C 255,0,0 -B 0,0,255
upper half: background black, red clock
footer half: completely blue

./clock -f --led-cols=64 --led-rows=32 ../fonts/7x13.bdf -d "%H:%M:%S"  -b 40 -C 255,0,0 -B 255,0,0
upper half: completely red
footer half: completely black

./clock -f --led-cols=64 --led-rows=32 ../fonts/7x13.bdf -d "%H:%M:%S"  -b 40 -C 255,0,0 -B 0,255,0
red clock upper half and full matrix green background

./clock -f --led-cols=64 --led-rows=32 ../fonts/7x13.bdf -d "%H:%M:%S"  -b 40 -C 255,0,0 -B 255,255,0
upper half: red clock and yellow background
footer half: green completely

**Comment by anp135** [comment]
I found this one: http://wiki.amperka.ru/продукты:rgb-led-matrix-driver-cap
here missed E and LAT and D both on the GPIO20. This is looks like adafruit-hat mapping instead strobe GPIO21. I made custom mapping from adafruit-hat and change .strobe from GPIO21 to GPIO20 - completely does not works, all matrix are black

**Comment by hzeller** [comment]
Maybe they have a typo in their documentation ? I suggest to mail the Amperka people directly.

**Comment by anp135** [comment]
Yes, looks like typo with strobe and should be GPIO21. Anyway can't access to blue in upper and red in footer halfs of matrix.

So you can't help me as I understood?

**Comment by hzeller** [comment]
It looks like a hardware issue with  your adapter board or  your matrix, so you have to follow the GPIO connections and see if they make it out of your board. And that your matrix boards are working. So nothing I can help you from here, this is a physical issue you have to check locally.

Of course, you should check also the troubleshooting section in the README, e.g. double checking that you don't have anything else running that accesses your GPIO pins.

**Comment by anp135** [comment]
thanks.

I checked with direct GPIO connection and raspberry perfect works with led panel directly. So this is hat hw problems.
thanks.

**Comment by alealexa** [comment]
> thanks.
> 
> I checked with direct GPIO connection and raspberry perfect works with led panel directly. So this is hat hw problems.
> thanks.

Please be aware that amperka uses -led-gpio-mapping=adafruit-hat-pwm mode.
Is also should be powered up directly from power supply to particular jack of hat.
Otherwise it will get some power from hub75 connection which enough to light "on" LED, but it wouldn't work properly.

Alexey

**Comment by anp135** [comment]
-led-gpio-mapping=adafruit-hat-pwm|adafruit-hat didn't help.
power supply connected to hat and also each panel connected to power directly.

**Comment by marcmerlin** [comment]
With apologies to everyone, this project is doing a one time cleanup of 
all the bugs that have not been addressed and closed in the last 10 years.

If you filed an issue to ask for help, please do not re-open this one or a new
bug, but go here instead: https://rpi-rgb-led-matrix.discourse.group/

Please be mindful that this project does not have any full time maintainers
or anyone paid. You should not expect personal help for your own projects or
answers to programming questions but you can ask for help with panels that
aren't working, setup questions, or anything that may be a bug here:
https://rpi-rgb-led-matrix.discourse.group/
If you need programming help and are willing to pay someone for it, please
state so in your message, and someone may be interested.

If you did report an actual bug that is still not fixed today, please accept
our apologies for this being mass closed, you can absolutely send a followup
and re-open the issue, or open a new up to date one as applicable.
Note that we very much welcome help, so if you can help fix your bug, you will
be a superstar :)

If your bug was linked to a pull request that was not resolved, please update
the pull request and the issue, and we will do our best to review them. We do
thank you for your undertanding that we have very limited free time, so any
extra work you do along with your bug report or PR request will be very much
appreciated and improve the odds that they will be looked at.

If you want to help contribute as a python or C# or any other language bindings,
you are also a superstar, please state so in the issue and we'll do our best to
give you the access and reviews you need.

For anything related to PWM panel support, please make sure you read README and
then refer to https://github.com/hzeller/rpi-rgb-led-matrix/issues/466 . 
Please do not comment if you're haven't read everything and/or if you're just
asking when your panel is going to be supported. Those bugs are if you're 
able to contribute to the effort.

For rPI5: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1603

For anything python, please link your bug to: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1828
For anything C#, please link your bug to: https://github.com/hzeller/rpi-rgb-led-matrix/issues/1828
